lib/puppet/util/tagging.rb

```diff
@@ -40,7 +40,7 @@ def tag(*ary)
   # since that results in testing the same string twice
   #
   def tag_if_valid(name)
-    if name.is_a?(String) && name =~ ValidTagRegex
+    if name.is_a?(String) && name =~ ValidTagRegex && Puppet::Resource::MAIN != name
```

Comment:
> If you think that tagging the main class with main is OK

TBH I'm not sure. For example, currently the `main` tag is exclude from `Stage[main]` and `Class[main]`, but `Class[settings]` is tagged with `settings`. I have no idea if that's intentional or just happens to work that way:

```json
    {
      "type": "Stage",
      "title": "main",
      "tags": [
        "stage"
      ],
      "exported": false
    },
    {
      "type": "Class",
      "title": "Settings",
      "tags": [
        "class",
        "settings"
      ],
      "exported": false
    },
    {
      "type": "Class",
      "title": "Main",
      "tags": [
        "class"
      ],
      "exported": false,
      "parameters": {
        "name": "main"
      }
    }
```